walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -126,27 +112,92 @@ private static boolean eventsOverlap(TimeRange currEventTimeRange,
+    if (startOfDayIsFree(eventList) && requestFits(request, firstEventTimeRange.start(), TimeRange.START_OF_DAY)) {
+      availableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY,
+        firstEventTimeRange.start(), false));
+    }
+
+    return availableTimes;
+  }
+  
+  /**
+   * Go through each event and check if it overlaps with another. If it does, we connect
+   * the end of our event with the start of the next one that is not overlapped. Otherwise,
+   * just connect the 2 events to create a time slot, if the request fits.
+   */
+  private static List<TimeRange> getMiddleOfDayTimeSlots(List<Event> eventList, MeetingRequest request) {
+    List<TimeRange> availableTimes = new ArrayList<>();
+    for (int i = 0; i < eventList.size() - 1; i++) {
+      TimeRange currEventTimeRange = eventList.get(i).getWhen();
+      TimeRange nextEventTimeRange = eventList.get(i + 1).getWhen();
+
+      if (eventsOverlap(currEventTimeRange, nextEventTimeRange)){
+        removeOverlappedEvents(i, eventList, currEventTimeRange); 
+      }
+
+      // The next event can be different from the one above, if it was overlapped
+      TimeRange actualNextEventTimeRange = eventList.get(i + 1).getWhen();
+      if (requestFits(request, actualNextEventTimeRange.start(), currEventTimeRange.end())) {
+        availableTimes.add(TimeRange.fromStartEnd(currEventTimeRange.end(),
+          nextEventTimeRange.start(), false));  
       }
     }
+
+    return availableTimes;
   }
 
   /**
-   * Check if our request fits between the end of current event and the start of next event.
+   * Method links the end of the last event to the end of the day.
+   * If there already are some available time slots added, we must check if nothing ends the day already
+   * If nothing was added yet, maybe our request fits only at the end of the day
    */
-  private static boolean requestFits(MeetingRequest request, int nextEventStart, int currEventEnd) {
-    return request.getDuration() <= nextEventStart - currEventEnd;
+  private static List<TimeRange> getEndOfDayTimeSlots(List<Event> eventList, MeetingRequest request) {
+    List<TimeRange> availableTimes = new ArrayList<>();
+
+    // Sort them again to be able to get the last end. Might be not that efficient.
+    Collections.sort(eventList, Event.ORDER_BY_END);
+    TimeRange lastEventTimeRange = eventList.get(eventList.size() - 1).getWhen();
+    
+    // Handle the case where the meeting can take place at the end of the day
+    if (!availableTimes.isEmpty() && requestFits(request, TimeRange.END_OF_DAY, lastEventTimeRange.end())) {
+      if (nothingEndsTheDay(availableTimes, eventList)) {
+        availableTimes.add(TimeRange.fromStartEnd(lastEventTimeRange.end(),
+          TimeRange.END_OF_DAY, true));
+      } 
+    } else if (fitsOnlyAtTheEnd(request, availableTimes, eventList)) {
+        availableTimes.add(TimeRange.fromStartEnd(lastEventTimeRange.end(),
+          TimeRange.END_OF_DAY, true));
+      }
+
```

Comment:
Only now I realized you have the following pattern:

```
if (A && B) {
  if (C) {
    action();
  }
} else if (D) {
  action();
}
```

That could be simplified as:

```
if ((A && B && C) || D) {
  action();
}
```